<center><h1>Introduction to Deep Learning & Neural Networks with Keras</h1></center>

 Build a regression model using the Keras library to model data about concrete compressive strength
 
The predictors in the data of concrete strength include:

- Cement
- Blast Furnace Slag
- Fly Ash
- Water
- Superplasticizer
- Coarse Aggregate
- Fine Aggregate

<h2><b>PART A - Build a baseline model </b></h2>

Download Data

In [1]:
import pandas as pd
import numpy as np

concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [2]:
concrete_data.shape

(1030, 9)

Prepare features

In [3]:
X = concrete_data
X = X.drop(columns='Strength', axis=1)
print("X.shape: ", X.shape)
X.head()

X.shape:  (1030, 8)


Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [4]:
Y=concrete_data['Strength']
print("Y.shape: ", Y.shape)
Y.head()

Y.shape:  (1030,)


0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64


Use the Keras library to build a neural network with the following:
- One hidden layer of 10 nodes, and a ReLU activation function
- Use the adam optimizer and the mean squared error as the loss function.

In [5]:
import keras
from keras.models import Sequential
from keras.layers import Dense

Using TensorFlow backend.


In [6]:
n_cols = X.shape[1]

def get_nn_regression_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))

    model.compile(optimizer='adam', loss="mean_squared_error")
    return model

model = get_nn_regression_model()

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.



In [7]:
from sklearn.model_selection import train_test_split

train_x, test_x, train_y, test_y = train_test_split(X,Y, test_size=.30)

print("train_x.shape: ", train_x.shape)
print("test_x.shape: ", test_x.shape)
print("train_y.shape: ", train_y.shape)
print("test_y.shape: ", test_y.shape)

train_x.shape:  (721, 8)
test_x.shape:  (309, 8)
train_y.shape:  (721,)
test_y.shape:  (309,)


2. Train the model on the training data using 50 epochs.



In [69]:
model.fit(train_x, train_y, validation_split=0.3, epochs=50)

Train on 504 samples, validate on 217 samples
Epoch 1/50
504/504 [==============================] - 15s 29ms/step - loss: 14289.1753 - val_loss: 7911.4465
Epoch 2/50
504/504 [==============================] - 0s 415us/step - loss: 5410.3388 - val_loss: 3293.0544
Epoch 3/50
504/504 [==============================] - 0s 319us/step - loss: 3084.8342 - val_loss: 2631.1640
Epoch 4/50
504/504 [==============================] - 0s 471us/step - loss: 2542.5940 - val_loss: 2089.9603
Epoch 5/50
504/504 [==============================] - 0s 353us/step - loss: 2015.4417 - val_loss: 1699.5105
Epoch 6/50
504/504 [==============================] - 0s 418us/step - loss: 1644.2032 - val_loss: 1417.1503
Epoch 7/50
504/504 [==============================] - 0s 378us/step - loss: 1336.8563 - val_loss: 1179.6257
Epoch 8/50
504/504 [==============================] - 0s 417us/step - loss: 1098.8520 - val_loss: 987.7656
Epoch 9/50
504/504 [==============================] - 0s 376us/step - loss: 916.0581 - val

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.



In [11]:
from sklearn.metrics import mean_squared_error

yhat = model.predict(test_x)
error_partA = mean_squared_error(test_y, yhat)
print('error : ' , error_partA)

error :  190.567049312125


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.



In [71]:
MSE_errors = []

for i in range(50):
    model = get_nn_regression_model()
    #split data
    train_x, test_x, train_y, test_y = train_test_split(X,Y, test_size=.30)
    #train model
    model.fit(train_x, train_y, validation_split=0.3, epochs=50, verbose=0)
    #evaluate model
    yhat = model.predict(test_x)
    error = mean_squared_error(test_y, yhat)
    MSE_errors.append(error)

5. Report the mean and the standard deviation of the mean squared errors.

In [73]:
print("MSE Errors - mean" , np.mean(MSE_errors))
print("MSE Errors - standard deviation" , np.std(MSE_errors))

MSE Errors - mean 546.275818282928
MSE Errors - standard deviation 801.1974919336554


<h2><b>PART B - Normalize the data</b></h2>

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [9]:
def normalize_inputs():
    X = concrete_data
    X = X.drop(columns='Strength', axis=1)
    X = (X - X.mean()) / X.std()  
    return X

In [75]:
X = normalize_inputs() 

MSE_errors = []

for i in range(50):
    model = get_nn_regression_model()
    #split data
    train_x, test_x, train_y, test_y = train_test_split(X,Y, test_size=.30)
    #train model
    model.fit(train_x, train_y, validation_split=0.3, epochs=50, verbose=0)
    #evaluate model
    yhat = model.predict(test_x)
    error = mean_squared_error(test_y, yhat)
    MSE_errors.append(error)

In [77]:
print("MSE Errors - mean" , np.mean(MSE_errors))
print("MSE Errors - standard deviation" , np.std(MSE_errors))

MSE Errors - mean 653.3811739812693
MSE Errors - standard deviation 135.34665461975644


<b>Compared to A, there is not much diference in the mean of the errors, but the standard deviation is smaller</b>

<h2><b>PART C - Increate the number of epochs</b></h2>

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [12]:
X = normalize_inputs()

MSE_errors_partC = []

for i in range(50):
    model = get_nn_regression_model()
    #split data
    train_x, test_x, train_y, test_y = train_test_split(X,Y, test_size=0.30)
    #train model
    model.fit(train_x, train_y, validation_split=0.3, epochs=100, verbose=0)
    #evaluate model
    yhat = model.predict(test_x)
    error = mean_squared_error(test_y, yhat)
    MSE_errors_partC.append(error)


In [14]:
print("MSE Errors - mean" , np.mean(MSE_errors_partC))
print("MSE Errors - standard deviation" , np.std(MSE_errors_partC))

MSE Errors - mean 213.66439857456174
MSE Errors - standard deviation 29.547836419368103


<b>Compared to B, both the mean and stanrd deviation of the errors are smaller</b>

<h2><b>PART D - Increase the number of hidden layers </b></h2>

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.



In [15]:
def get_deep_nn_regression_model():
    model_deep = Sequential()
    model_deep.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model_deep.add(Dense(10, activation='relu'))
    model_deep.add(Dense(10, activation='relu'))
    model_deep.add(Dense(1))

    model_deep.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [16]:
X = normalize_inputs()

train_x, test_x, train_y, test_y = train_test_split(X,Y, test_size=.30)

MSE_errors_partD = []

for i in range(50):
    model_deep = get_deep_nn_regression_model()
    #split data
    train_x, test_x, train_y, test_y = train_test_split(X,Y, test_size=.30)
    #train model
    model_deep.fit(train_x, train_y, validation_split=0.3, epochs=50, verbose=0)
    #evaluate model
    yhat = model_deep.predict(test_x)
    error = mean_squared_error(test_y, yhat)
    MSE_errors_partD.append(error)

In [18]:
print("MSE Errors - mean" , np.mean(MSE_errors_partD))
print("MSE Errors - standard deviation" , np.std(MSE_errors_partD))

MSE Errors - mean 52.061332353000395
MSE Errors - standard deviation 23.454265469783824


How does the mean of the mean squared errors compare to that from Step B?

<b>Compared to B, mean and standard deviation of errors are smaller</b>

<b>Compared to C, mean of errors is smaller, and standard deviation is similar</b>

<h2>Summary</h2>
A deeper model with normlized data obtains less error than a normalized model with more epochs

<b>PART A</b>
- MSE Errors - mean 546.275818282928
- MSE Errors - standard deviation 801.1974919336554

<b>PART B : Normalized data</b>
- MSE Errors - mean 653.3811739812693
- MSE Errors - standard deviation 135.34665461975644

<b>PART C : Normalized data + more epochs</b>
- MSE Errors - mean 213.66439857456174
- MSE Errors - standard deviation 29.547836419368103

<b>PART D : Normalized data + more hiden layers</b>
- MSE Errors - mean 52.061332353000395
- MSE Errors - standard deviation 23.454265469783824